<launch>
  <!-- Debug & Parameters -->
  <arg name="launch_prefix" default=""/> <!-- xterm -fs 60 -e gdb args -->
  <arg name="num_agents"/>
  <arg name="logging"       default="true"/>
  
  <node name="pose_graph_backend" pkg="pose_graph_backend" type="pose_graph_node" output="screen" launch-prefix="$(arg launch_prefix)">
    <param name="bow_voc" value="$(find pose_graph_backend)/conf/briskOKVIS_k10_L6_voc_v1.yml.gz"/>
    <param name="num_agents" value="$(arg num_agents)" />

    <param name="cam_config0" value="$(find pose_graph_backend)/conf/ncamera_23026.yaml"/> <!-- JOHNNY -->
    <rosparam param="gps_reference0"> [47.408624, 8.5044504, 561.459]</rosparam>
    <rosparam param="gps_offset0"> [0.0009, -0.0541, -0.1734]</rosparam>
   
    <!-- SYSTEM PERFORMANCE PARAMETERS -->
    <param name="gps_align_num_corr"              value="6"/>
    <param name="gps_align_cov_max"               value="0.025"/> 
    <param name="rel_pose_corr_min"               value="12"/>
    <param name="rel_pose_outlier_norm_min"       value="0.95"/>
    <param name="local_opt_window_size"           value="8"/> 

    <param name="loop_candidate_min_score"        value="0.02"/>
    <param name="max_loop_candidates"             value="3"/>
    <param name="loop_image_min_matches"          value="50"/> 
    <param name="loop_detect_sac_thresh"          value="25"/>
    <param name="loop_detect_sac_max_iter"        value="100"/> 
    <param name="loop_detect_min_sac_inliers"     value="9"/> 
    <param name="loop_detect_min_sac_inv_inliers" value="9"/> 
    <param name="loop_detect_min_pose_inliers"    value="9"/>

    <param name="loop_detect_reset_time"          value="6"/>
    <param name="loop_detect_skip_kf"             value="1"/> <!-- set to 1 for process every kf -->
    
    <param name="information_odom_drift_yaw"      value="20000"/>
    <param name="information_odom_drift_p"        value="5000"/>
    <param name="information_odom_map_yaw"        value="200"/>
    <param name="information_odom_map_p"          value="10"/>
    <param name="information_odom_edges_yaw"      value="100000"/>
    <param name="information_odom_edges_p"        value="10000"/>
    <param name="information_loop_edges_yaw"      value="2"/>
    <param name="information_loop_edges_p"        value="1"/>

    <param name="ignore_gps_altitude"             value="true"/>
    <param name="simulation"                      value="false"/>
    <param name="logging"                         value="$(arg logging)"/>
    <!-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ -->
    
    <!-- SYSTEM TOPIC REMAPPING - TODO Adjust here -->
    <param name="gps_active_0" value="true"/>
    <remap from="/keyframe0"  to="/pegasus/pcl_fusion_0/fused_vins_keyframe" /> 
    <remap from="/odometry0"  to="/pegasus/vins_estimator/odometry" />
    <remap from="/fake_gps0"  to="/pegasus/vrpn_client/raw_transform" />
    <remap from="/gps0"       to="/pegasus/gps" />
    <remap from="/fused_pcl0" to="/pegasus/pcl_fusion_0/fused_pcl"/>
  </node>

  <!-- RVIZ -->
  <node name="rvizvisualisation" pkg="rviz" type="rviz" output="log" args="-d $(find pose_graph_backend)/conf/pgbe_rviz.rviz" />
  
</launch>
